Intel cAVS v2.5 offline loader and multiprocessing

andyross opened the pull request:
Enable the MP cores on this hardware and make that configuration default.  Doesn't pass everything yet, but is a lot closer than what we have in tree.  Includes all relevant changes from the unmerged PR #36528

Commit:
soc/intel_adsp: Increase init priority of trace windows

These windows control host visibility of the trace output buffer.  The
buffer itself is writable memory always, but until we get to the
register init the host can't see them.  Since they contain
printk/logging output, they REALLY need to be initialized earlier than
anything else.

Also remove a rogue memset of the trace buffer.  That buffer is
already being initialized in a lazy-evaluated way by the trace output
code, and blowing it away here has the effect of forgetting anything
earlier code was trying to log!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Update cavs-fw.py loader for cAVS 2.5 devices

Add a loader script variant for Tiger Lake (cAVS 2.5) devices, which
have very slightly different loading behavior from older 1.5 DSPs.

This is added as a "-v25.py" script, and the original has been renamed
to cavs-fw-v15.py.  Note that there is no good reason except schedule
pressure that these are not the same script, I just wasn't able to
make a single script work compatibly in the time available.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Fix timing/clock register ownership on cAVS 1.8+

The wall clock timer is not (per documentation) part of the
"timestamping" register set on the DSP.  And its counter and
comparator registers work fine always.  But if the DSP isn't set as
the "owner" of the timestamp hardware, wall clock interrupts never
arrive.

Also grab the PLL ownership too, because SOF already does anyway.
While we don't have a dynamic clock driver yet, we will surely want
one soon and will needt this.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/intel_adsp: Update cavs-fw.py, robustify vs. Linux PM

Add support for the extended manifest blocks that recent versions of
rimage are including as a prefix on the firmware blob.

Also include some PCI runtime PM tricks to reliably enable the DSP
device on kernels that have turned off the DSP device automatically
(e.g. systems that have PM enabled by default but where the SOF driver
is not loaded -- chromebooks work this way, potentially other
distros).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/intel_adsp_cavs25: Add board documentation

The default development environment for this target is in fact a
consumer Chromebook available to anyone.  So it deserves public
documentation.

Note that this spends considerably more time explaining the details of
how to configure a chromebook as a Zephyr development platform than it
does the process for building and running Zephyr (which is really very
conventional).  Chromebooks, which allow user-signed audio firmware,
are a great boon to SOF development.  But they were never intended as
developer devices themselves.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

dcpleung committed:
xtensa: intel_adsp: align SoC initialization with SOF

This aligns the SoC initialization with the one in SOF,
especially the manipulation of clock control and power control
registers. These registers are not entirely the same across
CAVS versions, so we need to deal with them according to
which version we are building for. This also consolidates
the macros for these registers to the one provided by SOF
(soc/shim.h) to avoid duplication. Another note is that
the usage of clock gating bit was not correct. In SOF,
clock gating of SoC cores should be allowed but the old code
in Zephyr prevented clock gating, which has the potential to
prevent the whole DSP from going into low power mode.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

Commit:
soc/intel_adsp: Fix MP startup for cAVS 2.5

This hardware works a little differently.  The cores will start up
immediately on receipt of an IDC interrupt (they don't need the host
to be involved), but they don't have a ROM.  They start executing at
the start of the LP-SRAM block always.  Copy over a tiny trampoline
for them that jumps to the existing multiprocessor startup path.

Also set the PS WOE bit to enable register windows in the startup
path.  This isn't the hardware default, and where the ROM would do
that for us before here we need to make sure it's on.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Fix region cacheability for MP cores

On MP cores that don't come through the core entry point
(e.g. TGL/v2.5) we reach C code with hardware defaults for the RPO/TLB
settings.  Set these up correctly on entry.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Fix ATOMCTL on MP startup

Hardware defaults for the secondary CPUs have the S32C1I instruction
set to be atomic only with respect to the local L1 cache, which is
basically useless on a multiprocessor platform.  The CPU0 boot path
sets this manually, so we need to duplicate that here.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Add full cache enable logic

Earlier platforms were relying on the system ROM to have done this
correctly, but with CAVS 2.5 we launch the CPU into our own code
directly.  So we need to do those steps manually.  And there's also a
new one on this hardware, which has software power control over the
cache SRAM.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: New IDC driver

The original interface for the intra-DSP communication hardware on
these devices was buried inside a Zephyr IPM implementation.
Unfortunately IPM is a two-endpoint point-to-point communication
layer, it can't represent the idea of devices with more than 2 cores.
And our usage (to push a no-argument/no-response scheduler IPI) was
sort of an abuse of that metaphor anyway.

Add a new IDC interface at the SOC layer, borrowing the C struct
convention already used for the DSP shim registers.

Augment with extensive documentation, extracted via a ton of
experimentation on cAVS 2.5 hardware.

Note that this leaves the previous driver in place for the cavs_v15
and intel_s1000 devices.  In principle they should use it too (the
hardware registers are identical), but this hasn't been validated yet.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
Revert "soc: intel_adsp: fix linker script for XCC"

This reverts commit ee7773fb462bbd600668aac12cb41eee2fbc3535.

Unfortunately this mechanism doesn't seem to actually work on the SDK
linker.  The emitted sections, when passed a symbol name as the "start
address" just appear wherever the "." variable was pointing (in this
case, into the cached region).  That breaks the kernel coherence
layer, obviously.

Revert for now, which will regress the XCC build fix pending a proper
root cause.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: cavs_v25: Clean up platform config to reflect recent work

Use the built-in IDC handling and not IPM (which is limited to two
CPUs).  Declare two cpus for now, Zephyr tests are having problems
with more at the moment (that isn't a CI configuration, so we may have
work to do).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Note that this includes a regression of the XCC linker fix from @dcpleung, which unfortunately seems to misbehave on SDK builds (my guess is that the syntax using a symbol name as a section start address gets ignored?).  So we're likely going to need to revisit that bug again and find a fix that works identically to the historical code.

Commit:
soc/intel_adsp: Keep track of started CPUs in the SOC layer

On cAVS 2.5, there is an inherent race with the IDC interrupt.  It's
used for routine IPIs during OS operation, but also for launching a
power-gated core.  Recent changes moved the unmasking of the IDC
interrupt earlier, which made it possible for early OS scheduler
behavior (e.g. adding the main thread to the run queue) to
accidentally launch the other cores into LP-SRAM that had not been
initialized.

Instead of treating this with initialization ordering, keep and
maintain a list of active CPUs and check them at runtime to be sure we
never try to IPI a CPU that isn't running yet.  We're going to need
this feature when we add live core offlining anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/xtensa: Fix outgoing stack flush for dummy threads

On CPU startup, When we reach the cache flush code in arch_switch(),
the outgoing thread is a dummy.  The behavior of the existing code was
to leave the existing value in the SR unchanged (probably NULL at
startup).  Then the context switch would walk from that address up to
the top of the outgoing stack, flushing everything in between.  That's
wrong, because the outgoing stack is a real pointer (generally the
interrupt stack of the current CPU), and we're flushing everything in
memory underneath it.

This also reverts commit 29abc8adc025 ("xtensa: fix booting secondary
cores on the dummy thread"), which appears to have been an early
attempt to address this issue.  It worked (modulo all the extra and
potentially incorrect flushing) on cavs v1.5/1.8 because of the way
the entry code worked there.  But on 2.5 we now hit the first context
switch in a case where those extra lines are in address space already
marked unwritable by the CPU, so the flush explodes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Add INTCTRL register interface

Add a struct-based interrupt masking API to match the existing shim
and IDC register interfaces.  The existing interrupt controller code
isn't using it yet.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/xtensa: Add arch_cpu_idle() workarounds

A simple WAITI isn't sufficient in all cases.  The cAVS 2.5 hardware
uses WAITI as the entry state for per-core power gating, which is very
difficult to debug.  Provide a fallback that simply spins in the idle
loop waiting for interrupts to provide a stable system while this
feature stabilizes.

Also, the SOF code for those platforms references a known bug with the
Xtensa LX6 core IP (or at least some versions), and will prefix the
WAIT instruction with 128 NOP.N's followed by an ISYNC and EXTW.  This
bug hasn't been seen under Zephyr yet, and details are sketchy.  But
the code is simply enough to import and works correctly.

Place both workaround under new kconfig variables and select them both
(even though they're actually mutually exclusive -- if you select both
CPU_IDLE_SPIN overrides) for cavs_v25.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Push a few fixes, see final three patches (one of which regresses the use of WAITI in Zephyr idle and is really only a workaround).

At this stage I can get passing results from basically all 1-cpu and single-feature tests, and most threading/smp tests in 2-core mode.  Some of that may be latent bugs in the tests: we haven't had a 4-core configuration in CI yet and so testing that many has only been done ad hoc.

Commit:
soc/intel_adsp/cavs_v25: Correct power gating state handling

There is a hardware startup state where power gating can be "enabled"
even though the core is actually launchable via an IDC interrupt (in
fact that's the hardware default).  In that state, the CPU will launch
correctly but then unexpectedly shut itself off then it enters the
idle thread.

Don't rely on initialization state, always set the power and clock
gating bits (to disable gating) immediately before CPU launch.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lgirdwood commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -117,46 +117,102 @@ BUILD_ASSERT(XCHAL_EXCM_LEVEL == 5);
+	/* Prefetcher also power gates, same interface */
+	CxL1PCFG = 1;
+	while ((CxL1PCFG & 0x10000) == 0) {
+	}
```
may need timeouts and errors here to cover any IP misconfiguration on host side or from other FWs

andyross replied:
Here I think there's less value.  This is REALLY early code, at this stage where the cache isn't enabled we can't communicate with other cores through memory, we can't take spinlocks, etc... So basically we can't even log an error if we detect one.

lgirdwood commented on soc/xtensa/intel_adsp/common/include/cavs-idc.h:
```diff
@@ -0,0 +1,84 @@
+		uint32_t tefc; /*  ^^ + (E)xtension       */
+		uint32_t itc;  /* (I)nitiator (T)o (C)ore */
+		uint32_t ietc; /*  ^^ + (E)xtension       */
+	} core[4];
```
Do we want to make this ```array[CONFIG_CORE_COUNT]``` ?

andyross replied:
Yes, but then we'd need variable-length padding after the block.  The register layout doesn't change based on how many cores are instantiated.  So it's simplest this way IMHO.

lgirdwood reviewed (approved):
I have a few minor comments, but feel free to disregard or fix incrementally if you care. Looks like we are firing on all 4 cylinders now !

Commit:
soc: intel_adsp/cavs_v25: Add CPU halt and relaunch APIs

Add a SOC API to allow for application control over deep idle power
states.  Note that the hardware idle entry happens out of the WAITI
instruction, so the application has to be responsibile for ensuring
the CPU to be halted actually reaches idle deterministically.  Lots of
warnings in the docs to this effect.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

dcpleung commented:
Maybe move `boards/xtensa/intel_adsp_cavs15/tools/cavs-fw-v25.py` under `boards/xtensa/intel_adsp_cavs25/tools`?

andyross commented on soc/xtensa/intel_adsp/common/soc.c:
```diff
@@ -195,34 +196,74 @@ static inline void soc_set_power_and_clock(void)
+	dsp_shim_regs->clkctl |= SHIM_CLKCTL_RHROSCC;
+	while ((dsp_shim_regs->clkctl & SHIM_CLKCTL_RHROSCC) !=
+	       SHIM_CLKCTL_RHROSCC) {
+		k_busy_wait(10);
```
Yeah, that seems reasonable.  Will add.

andyross commented on soc/xtensa/intel_adsp/cavs_v15/linker.ld:
```diff
@@ -35,16 +35,16 @@ PROVIDE(_MemErrorHandler = 0x00000000);
- * pointer forward or backward so code and data land in correct
- * region. Remember to align the memory, and be sure to also emit the
+ * These macros will set up a segment start address correctly,
+ * including alignment to a cache line.  Be sure to also emit the
```
The problem is I don't think we've identified a solution that duplicates the current behavior with the xcc linker yet; what we had was generating cached addresses always (on current binutils, I don't have xcc set up locally to be sure it wasn't doing the right thing there) so only works in uniprocessor setups.  But yes, the gold standard here is a cleanly abstracted linker hack.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -187,9 +188,39 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+	 */
+	void **lpsram = z_soc_uncached_ptr((void *)LP_SRAM_BASE);
+	uint8_t tramp[] = {
+		0x06, 0x01, 0x00, /* J <PC+8>  (jump to L32R) */
```
It's actually the assembler itself; it won't generate a L32R instruction unless you use an assembler generated immediate, which we can't do here because we don't link in lpsram at all.  There's a comment to that effect: what we really want is to have lp-sram be part of the same set of address spaces and visible for linker symbols, and then be copied out of the IMR bootstrap image in the same way.

FWIW: that sequence is really just a jump.  It's the smallest possible "jump to absolute address from anywhere" possible in Xtensa, hilariously.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -120,7 +120,15 @@ int cavs_idc_smp_init(const struct device *dev);
+	 * the shared bus
+	 */
+	reg = 0x15;
+	__asm__ volatile("wsr %0, ATOMCTL" :: "r"(reg));
```
Yeah, nothing is going to work if it's not there.  I stumbled on it because the new MP startup doesn't go through the whole main CPU initialization path and so short-circuited some very old Cadence-submitted code in our arch layer.  One of a few spots where the devices are "broken by default", ugh.

andyross commented:
> Maybe move boards/xtensa/intel_adsp_cavs15/tools/cavs-fw-v25.py under boards/xtensa/intel_adsp_cavs25/tools?

Actually the second I have a chance to do some housekeeping I think we want to (1) unify both scripts so they work on all hardware and (2) move it to soc/xtensa/intel_adsp and out of the board layer.  Likewise for a bunch of the headers, which differ only in a few addresses that are extractable from devicetree (some of that work is already in this PR).  Likewise for the linker scripts which are also almost all duplicated...

Really the "board" layer for these devices is way too thick.

andyross commented:
Updated with current version that is actually passing many tests now, and has a seemingly-working low power idle API (`soc_{halt,relaunch}_cpu(id);`) in the SOC layer (though no test for that yet -- have to clean up my hello_world-based rig and add it).

Commit:
scripts/check_compliance.py: Add kconfig false positives

New docs for cavs_v25 describe building a Linux kernel for a
Chromebook and need to talk about these kconfigs.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
The checkpatch whine is a false positive.  Pre-existing formatting for that inline assembly had those spaces, and objectively the current format is more readable than any alternatives.  The warning is supposed to be about trailing spaces in log messages or whatever,  it just doesn't apply.

lyakh reviewed (dismissed):
multiple build failures on APL / cAVR 1.5

Commit:
soc: intel_adsp_cavs25: Fix linker section overlap, support !COHERENCE

The alignment on .bss was coming out wrong.  The ". = ALIGN(4096);"
statement was being ignored, somewhat inexplicably.  This resulted in
the bss symbols being assigned corret-seeming, non-overlapping
addresses.  But it overlapped the page-sized padding at the end of
.data.

As it turns out, the rimage format (not the linker or Zephyr) requires
page-sized sections to copy, and the bootloader code does that copy by
writing to the CACHED mapping of the memory (.bss is, like .data,
uncached/coherent by default).  So at runtime the CPU was running in a
context where the cache was populated with "booby trap" data at the
start of .bss.  True .bss access would hit the memory uncached and see
the "correct" value, but at arbitrary times during execution lines
would be flushed out of L1 cache on top of it.

Oops.  This was found by accident, actually, as routine changes to the
linker script to correctly support the case where KERNEL_COHERENCE=n
(i.e. put everything in the cached mapping and nothing in uncached)
suddenly hit rimage failures because of the overlap.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Fix IDC masking & state issues

Fix various bugs with the new IDC layer that show up in edge cases
where code relies on correct timing of IPIs (unsurprisingly there is a
lot of code that recovers anyway even if the IPI doesn't arrive
promptly).  Leaving this as a separate patch because the prior code in
the PR has already been reviewed and it "mostly" worked:

The unmasking of the L2 interrupt bit (remember there are three layers
of masking of the IDC interrupt) was always operating on CPU0 at CPU
startup because the code had been copied blindly.  Unmask the CPU
we're actually launching.  It turns out cAVS 2.x re-masks this on CPU
launch automatically.

The global init code to unmask all these interrupts at startup had the
same bug, even though it turned out to be needless (the initialization
state has it unmasked until it turns it back off).  Do it right
anyway.  Similarly add code to clear out existing interrupt latch
state by ACKing all IDC interrupts at startup.  Seems needless, but
behavior isn't documented so let's be safe.

Flag CPU0 as always "active" for the purposes of IPIs.  Forgot to do
this earlier, oops.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
One last patch in the series (left unsquashed for easier review) seems to have been it.  I haven't had time to run a full suite yet, but I can't find any common test cases that fail.

Build failures on older cAVS should be fixed, but I don't have the board with me to test it right now.

lyakh commented:
Concerning the APL SMP runtime breakage: there are several changes to SMP code on platforms other than cAVS 2.5 in this PR, I cannot say which of those changes breaks APL. I could bisect, but first I'd need to merge the last fixing commit with the commit, that it's fixing, and then bisect. Maybe it's better to extract a minimum PR from this one, touching *only* 2.5 and not modifying a single instruction on others. After that we can improve and consolidate step by step.

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -312,28 +314,21 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+	uint32_t idc_reg = idc_read(IPC_IDCCTL, cpu_num);
+
 	idc_reg &= ~IPC_IDCCTL_IDCTBIE(0);
 	idc_write(IPC_IDCCTL, cpu_num, idc_reg);
```
is this the desired approach now, that sometimes IDC registers are accessed via `idc_read()` and `idc_write()` and sometimes via `IDC[src_core].core[dst_core].x` or is there a well defined rule when to use which approach?

andyross replied:
The rule is "use the new API", but I've left the older usage in place in a bunch of places out of conservatism.  In this case the delta is just a noop -- I removed the declaration of the same variable above this line so needed to add the "uint32_t " here.

andyross commented on kernel/smp.c:
```diff
@@ -66,8 +66,7 @@ void z_smp_thread_swap(void)
 
-#ifndef CONFIG_SMP_BOOT_DELAY
-static FUNC_NORETURN void smp_init_top(void *arg)
+static inline FUNC_NORETURN void smp_init_top(void *arg)
```
The use of "inline" in Zephyr-idiomatic GCC actually has nothing to do with inlining[1].  Instead, "static inline" is the technique we use where we have a symbol in a file that may or may not be referenced by other code in the same file, we're getting an "unused static" build failure from -Werror, and we don't want to continually repeat the same #ifdef block for all the permutations.  In this case I added a hook z_smp_start_cpu() that is called out of soc_relaunch_cpu() and isn't related to SMP_BOOT_DELAY.

[1] It doesn't actually do anything useful in practice.  The compiler is smart enough to inline forward-defined statics already, and in the rare cases where you actually do want to specifically tell the compiler to inline something, "inline" sometimes doesn't!  We have an ALWAYS_INLINE macro defined around the relevant attribute to control that for performance or correctness reasons.

Commit:
soc: intel_adsp: Unbreak older cAVS devices

On pre-2.5 cAVS, the initial IDC interrupt to start the other core is
handled by software in the firmware ROM.  That means that it has to be
unmasked for the mechanism to work (with 2.5, the interrupt is handled
by hardware regardless of what the masking state in the interrupt
controller is).

Similarly, the Xtensa Region Protection Option entries have already
been set by ROM code when we arrive in enable_l1_cache(), so we can
skip that part on older machines.  Also removed because trying to
rewrite those entries was causing inexplicable hangs on cAVS 1.5,
plausibly because the region had active cache lines.

(This patch is separate for easier review in a long evolving PR.
Technically it represents a bisection problem as the "New IDC Driver"
patch before this was a regression.  Seems like a safe enough thing to
handle if you land on this.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Add a patch that fixes an SMP regression on APL.  One part is straightforward: I accidentally removed code that unmasked the IDC interrupt on the target core.  That's benign on 2.5 where the startup-on-IDC behavior is handled by a state machine outside the interrupt controller, but on earlier hardware that's a software handler in ROM!

The other bit is sort of voodoo: the new scheme sets up the RPO memory mapping "TLB" registers, which is required on 2.5 where we handle the hardware handoff ourselves.  But trying to re-write them to the same values already set by the ROM is causing hangs.  We never used to do this, and it's needless since the ROM already does it, so remove.  But I don't know why that shouldn't work.  Maybe the cache controller (which has active lines for the region) is messing things up?

andyross commented:
And I see that pylint warning about duplicate lines is back.  @nashif you mentioned there was a way to add a comment to suppress that?  Any pointers would be appreciated.

andyross commented:
Simple cleanup and an attempt to suppress that pylint duplicate-code error with a pragma comment.

andyross commented:
Nope.  Turns out that cross-file pylint warnings can't be disabled (https://github.com/PyCQA/pylint/issues/214).

That makes that issue another genuine false positive.  AFAICT there's nothing that can be fixed with the compliance results.  Pull the trigger and merge, please.

andyross commented:
Had to respin to fix a warning that snuck in.

andyross commented:
Once more because I forgot to commit both warning fixes, ugh.

Commit:
samples/net/sockets: Exclude intel_adsp boards

These are getting build failures due to some kind of devicetree
console definition missing.  These devices don't do networking, and
strictly don't even have a proper console device (logging is done via
a host mechanism).  Probably fixable.  Not worth the trouble.  Filter.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Websocket samples?!  Ugh.

andyross commented:
Buildkite seems to have two remaining failures on the adp_xc7k_ae350 board.  I can guarantee that's unrelated.

lyakh commented:
Still not working on APL with multicore

andyross commented:
@lyakh Can you be more specific?  The test suite seems fine.  Really, we need to get this merged.  We can work bugs as they appear.

Commit:
soc: intel_adsp: Handle reset better in adsplog.py

When working with the SOF kernel driver, it likes to shut down the DSP
on error.  That means there is a very small window in which to catch
any log output, even with a whiteboxed kernel.  So we should be
polling much faster (10 Hz) for changes when we detect a reset.

Also, don't repeatedly log the device reset detection, it spams the
console badly when we crank up the rate.  Just log it once and then
stay silent until we see output.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Unify "active CPUs" state

This feature got written twice for two different purposes (to inform
the SOF app of which CPUs are running, and to predicate the delivery
of IPIs to the cores ready to receive the interrupt).  Use only one.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -354,12 +362,11 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 {
 #ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V25
-	uint32_t prid;
+	uint32_t curr = prid();
```
this seems to break bisect: `prid()` is only defined in a later commit https://github.com/zephyrproject-rtos/zephyr/pull/37647/commits/1eb62d4fc1ffd4ff2ae1528243d7ae6f2d9cd0c1

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -447,3 +458,81 @@ void soc_idc_init(void)
+		goto out;
+	}
+
+	CAVS_INTCTRL->l2.clear = CAVS_L2_IDC;
```
this should also be `CAVS_INTCTRL[...].l2.clear` I guess?

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -359,13 +351,80 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+	for (int c = 0; c < CONFIG_MP_NUM_CPUS; c++) {
+		if (c != prid) {
+			IDC[prid].core[c].itc = BIT(31);
+		}
```
We never really had an IDC driver.  We had an IPM driver that managed the IDC device, but IPM is a two-endpoint communication API and won't work for more than 2 cores.  And yes, it's going away.  2.5 uses the new interface exclusively, and I'll remove support (and the ipm_cavs_idc driver)  after the next release.

Effectively, soc_mp.c is now the "IDC driver", which makes sense to me.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -247,8 +247,12 @@ void z_mp_entry(void)
+	 * on older hardware or by the hardware itself after delivery
+	 * of the startup IDC.  We're ready for it now.
+	 */
+	CAVS_INTCTRL[start_rec.cpu].l2.clear = CAVS_L2_IDC;
```
Wait, this one line?  But this is critical, without it the new CPU won't be receiving IDC interrupts, which are how the scheduler informs other CPUs of state changes that might preempt their running thread.

lyakh replied:
yes, just this one line - this is the diff to my fork. And @kv2019i also reports that my fork is working for him on TGL multicore

lyakh replied:
sorry, by "1 line" I meant 1 removed and 1 added line of code - that your hunk, that removed `irq_enable()` and replaced it with the `CAVS_INTCTRL[]...` assignment. I presume, your assignment only enables the IDC interrupt on Level 2 IRQ controller, whereas the original line was also enabling the Level 2 interrupts on the core.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -120,7 +120,15 @@ int cavs_idc_smp_init(const struct device *dev);
+	 * the shared bus
+	 */
+	reg = 0x15;
+	__asm__ volatile("wsr %0, ATOMCTL" :: "r"(reg));
```
Yes, the ROM sets this on pre-2.x devices.  It never worked on 2.5, which is why it had to be added in software.  SOF has the same thing in the "alternate reset vector" startup assembly.

Commit:
soc: intel_adsp: Lengthen initialization sleep on the cavs 1.5 loader

This was still observed to be failing occasionally.  Seems like 100ms
is more robust than 10ms.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -337,6 +340,9 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	z_mp_stack_top = Z_THREAD_STACK_BUFFER(stack) + sz;
 
+	/* Pre-2.x cAVS delivers the IDC to ROM code, so unmask it */
+	CAVS_INTCTRL[1].l2.clear = CAVS_L2_IDC;
```
IIRC, this is there for the halt/relaunch API, which masks IDCs to prevent races.

Commit:
soc: intel_adsp: Abstract out a prid() accessor

Useful utility.  Should probably have this in the arch layer.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
OK, indeed @lyakh found the magic.  That irq_enable() was also setting the Xtensa INTENABLE SR, not just the second level interrupt controller.  So the interrupt would be able to reach the core but would still be masked by the CPU itself.  Under Zephyr-internal code, that was hidden by the fact that the timer driver initialization on the second core (which shares the same L2 interrupt) was then enabling it before any user code ran.

But on SOF, secondary core initialization happens later, and the IPC task run immediately after core start needs to be able to run synchronously **on that core**, and can't wait for a timer setup.  Instead it was being picked up on core 0 and failing.  (Which is sort of a SOF bug, actually -- if the task has to run on a specific core for correctness it should be pinned to run only on that core, can fix later.)

andyross commented:
Also added a few new minor patches from yesterday's work and fixed the identified bisectability/misordering problems.  Probably should have rebased over main, but didn't.  Will do if it needs a respin.  @lyakh, please review and retest?

andyross commented:
Unsurprisingly I indeed had to respin to add a dropped patch.  So now it's rebased on current main.

kv2019i reviewed (approved):
Good now, tested on cavs25/TGL. I still get an error when pipeline is teared down, but the DSP is powered down anyways, so I think this is something we can tackle incrementally.

lyakh commented:
(lightly) tested to work with SOF on APL with multicore topology again - thanks!

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -122,6 +122,23 @@ void z_mp_entry(void)
+	for (int region = 0; region < 8; region++) {
+		uint32_t addr = 0x20000000 * region;
+
+		__asm__ volatile("wdtlb %0, %1" :: "r"(attribs[region]), "r"(addr));
```
It is (was!  finally merged!) an almost three month old PR, that's just going to happen. :)   What happened was that the code just got moved.  It was added in the original spot because it's affirmatively needed for 2.5 (because with no ROM nor architecture assembly, nothing else does it).  Then everything was unified to a single core startup path in a later patch because that's a cleaner design.  You're right that it could have all been done at once, but that would have led to much larger patches and/or a ton of splitting and rework, and I didn't think this was so strange to look at.  It only seems weird when you're looking at it in a single PR.  If you look at the commits in the tree it's a natural evolution.
